### DISCUSSION TITLE
Immich HW acceleration

### DISCUSSION THREAD

**Discussion by di5cord20** [tteck/Proxmox#1943] [Help Needed] [answered]
I created a Docker LXC and installed Immich.  I'd like to leverage HW Acceleration but the /dev/dri path doesn't exist on the LXC.  It does for the Jellyfin LXC I previously created (both with your scripts).

Do I need to do any additional steps to expose/include the /dev/dri path to the Docker LXC?

![image](https://github.com/tteck/Proxmox/assets/140566995/de46a4c5-c1fb-46c5-8015-75b9dc6fd1bf)

**Comment by tteck** [comment]
Start here
https://github.com/tteck/Proxmox/blob/main/misc/build.func#L550

**Reply by di5cord20** [reply]
Thanks @tteck .  Do I just run those lines in my LXC shell to mount and configure the paths, groups and permissions?

**Reply by tteck** [reply]
No. Within the Proxmox shell execute `nano /etc/pve/lxc/111.conf` replacing `111` with your container ID. then add:
```
lxc.cgroup2.devices.allow: c 226:0 rwm
lxc.cgroup2.devices.allow: c 226:128 rwm
lxc.cgroup2.devices.allow: c 29:0 rwm
lxc.mount.entry: /dev/fb0 dev/fb0 none bind,optional,create=file
lxc.mount.entry: /dev/dri dev/dri none bind,optional,create=dir
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file
```
Reboot the container to apply the new settings.
Note: The container needs to be Privileged.

**Reply by di5cord20** [reply]
Also, as a suggestion, would it make sense to enhance the advanced install function to ask if HW acceleration should be added?  I realize it's a niche use case so may not make sense or cause confusion.

As an alternative, perhaps a script on its own where the user specifies the container and the script does its magic.

**Reply by tteck** [reply]
> As an alternative, perhaps a script on its own where the user specifies the container and the script does its magic.

Possible

**Reply by di5cord20** [reply]
Any way to change an Unprivileged to Privileged?  Or will I need to start from scratch?

**Reply by tteck** [reply]
Start over

**Reply by di5cord20** [reply]
Thanks for your help.  Very clear and straightforward.

Last question, how can I confirm if a container is Privileged post install?  I had challenges with HW acceleration on Jellyfin so disabled it through the app.  I now wonder if potentially I didn't set it as Privileged.  I did verify that the /dev/dri does exist though.

**Reply by tteck** [reply]
![Screenshot 2023-10-20 7 32 11 AM](https://github.com/tteck/Proxmox/assets/21342670/29383b52-aeef-42ed-ba2f-fe2a6e1b1c33)

**Reply by di5cord20** [reply]
+1 more suggestion.  For the apps your script checks for to include the HW acceleration settings in the container conf, should a note be added to the app description instructing the user that the container must be Privileged?

**Reply by tteck** [reply]
Any app that installs HW acceleration is set to Privileged by default.

**Reply by di5cord20** [reply]
Perhaps that last suggestion isn't needed.  I just validated both my Jellyfin and Plex containers are Privileged.  I can't recall if I set it manually or if the script did so automatically. It's been a while, but perhaps your install script even instructed me to set as Privileged for those apps which would be perfect.

Thanks again for your fantastic scripts and support.

**Reply by tteck** [reply]
example, Jellyfin is Privileged by default
https://github.com/tteck/Proxmox/blob/108c4e1a14de47c73f306457dc274db56645800d/ct/jellyfin.sh#L33

whereas Adguard Home is not
https://github.com/tteck/Proxmox/blob/108c4e1a14de47c73f306457dc274db56645800d/ct/adguard.sh#L33

**Comment by tteck** [comment, marked as answer]
@di5cord20, run the command below in the Proxmox VE Shell to add HW Acceleration to an existing LXC.

```
bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/misc/hw-acceleration.sh)"
```

**Reply by di5cord20** [reply]
Thanks for the script.  I assume it will come in handy for many.  I was able to manually update Immich per your previous instructions.  I'll try the script you provided now this weekend just to check it out.

**Reply by tteck** [reply]
The script installs all of the necessary packages also.

**Reply by di5cord20** [reply]
I ran the script.  It instantly found all containers that were privilaged and let me select the immich container.  It proceeded to update and then instructed me to reboot the lxc container once completed.  I then ran ```nano /etc/pve/lxc/111.conf``` (changing 111 to my container id) and saw the added lines.  That said, I had previously manually added the lines per your instrutions (prior to the script being available) so I ended up deleting the lines I had previously manually added.

I can see this script being helpful for others in this scenario.  It would be great if the script could detect if the required lines are already added vs. duplicating them (but an unlikely scenario if somebody is using the script).

**Reply by rcunn87** [reply]
Sorry to interject but what are the scripts doing that requires the container to be privileged for hw-acceleration? Before I found these scripts I was able to get my unprivileged plex container to run quicksync just fine.

**Reply by drzett** [reply]
@tteck Hi, first of all, I would like to thank you for saving me a lot of work through your efforts. Here's a tiny note about the script: cat adds the mapping for /dev/dri at the 'end of file' to the <ID>.conf. This doesn't work if you've created a snapshot for the LXC, as the second part of the .conf in that scenario refers to the parent of the snapshot, not the current version.

Something like replacing the 'cat' block with this, should work in any case:

```
entries_to_add='lxc.cgroup2.devices.allow: c 226:0 rwm\
lxc.cgroup2.devices.allow: c 226:128 rwm\
lxc.cgroup2.devices.allow: c 29:0 rwm\
lxc.mount.entry: /dev/fb0 dev/fb0 none bind,optional,create=file\
lxc.mount.entry: /dev/dri dev/dri none bind,optional,create=dir\
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file'

# Add (only) after first instance/line of "lxc.cap.drop:"
sed -i "/lxc.cap.drop:/a $entries_to_add" "/etc/pve/lxc/${privileged_container}.conf"
```

If I'm right, the LXC also has to restart after running this to make the new dri mount available. To run the (full) script twice should do the trick. (Run it once -> shutdown/start -> run again -> restart, done)

**Comment by mastan30** [comment]
Here's a way to do HW acceleration for unprivileged containers:

# Immich Docker Install

1. To install immich, the recommended way to install is using a docker compose, we can't install it on a bare metal debian os. So we will install a docker LXC using the proxmox script and install docker compose.Then we will create the docker-compose file for installing immich.

2. To install docker LXC we can used the following Proxmox script from the beloved tteck github:

		

		Docker LXC
		Options to Install Portainer and/or Docker Compose V2
		Docker is an open-source project for automating the deployment of applications as portable, self-sufficient containers.

		🛈 If the LXC is created Privileged, the script will automatically set up USB passthrough.

		To create a new Proxmox VE Docker LXC, run the command below in the Proxmox VE Shell.

		Copy
		bash -c "$(wget -qLO - https://github.com/tteck/Proxmox/raw/main/ct/docker.sh)"
		⚡ Default Settings: 2GB RAM - 4GB Storage - 2vCPU ⚡
		As an alternative option, you can use Alpine Linux and the Docker package to create a Docker LXC container with faster creation time and minimal system resource usage.

		To create a new Proxmox VE Alpine-Docker LXC, run the command below in the Proxmox VE Shell.

		Copy
		bash -c "$(wget -qO - https://github.com/tteck/Proxmox/raw/main/ct/alpine-docker.sh)"
		⚡ Default Settings: 1GB RAM - 2GB Storage - 1vCPU ⚡
		⚠ Run Compose V2 by replacing the hyphen (-) with a space, using docker compose, instead of docker-compose.

		Portainer Interface: IP:9443


3. Once the container is installed, let's mount a mountpoint of SMB share where the images are uploaded.

4. In order to write the files into the SMB share, there's a catch here. The LXC we created using the script is an unpriviliged container which is the safest way to build container but the issue is we can't write anything into the shared drived from Proxmox host.

5. Instead of mounting the SMB share via Proxmox UI, we use fstab to mount the SMB share. Here are the steps to mount

	```
	  vi /etc/fstab
	```
	
	Then add the following line replacing the uid and gid for which you want the access to the proxmox container
	
	```
	  //192.168.X.XX/Path/to/share/ /path/to/mount/       cifs    noperm,iocharset=utf8,rw,credentials=/root/.storage_credentials,uid=1000,gid=1000,file_mode=0660,dir_mode=0770 0 0
	```
     eg: //192.168.1.19/Photos/immichtest/ /mnt/pve/immichtest/       cifs    noperm,iocharset=utf8,rw,credentials=/root/.storage_credentials,uid=1000,gid=1000,file_mode=0660,dir_mode=0770 0 0
	 
	```
      vi /root/.storage_credentials
    ```	
	 In this file add the username and password as below:
	 
	```
	username=username
	password=password
	```
	
	```
	mount -a
	```
   This will mount the SMB share into the Proxmox host, now check the mount directory permissions by doing ls -al and check if the mount has the permissions of assigned uid and gid. If not do a reboot of the host and it takes care of it.
   
6. Shutdown the container and mount the SMB share to the proxmox LXC by following commands:

   ```
    pct set 100 -mp0 /mnt/bindmounts/shared,mp=/shared
   
   ```
    eg: pct set 100 -mp0 /mnt/pve/immichtest/ ,mp=/immichtest/   - 100 is the lxc id
   
7. Start the container and check the if the mount point is assigned with right uid and gid inside the container.

8. Now in the LXC create a folder using following commands

   ```
     mkdir ./immich/
	 cd ./immich/
   ```
   
   create the docker-compose file and also the .env file in the directory as per the immich guide
   
   ```
      https://immich.app/docs/install/docker-compose
   ```
   
   Run the docker compose and follow the immichi guide to setup your albums and libraries as required.
   
   
9. Optinally in order to add hardware acceleration, I used NVIDIA acceleration. In order to get it working here are the steps to follow:
   
   Next step is to passthrough the nvidia graphics card to the lxc, In order to do the passthrough follow the below guide.
   
   > [Nvidia graphics passthrough](https://gist.github.com/egg82/90164a31db6b71d36fa4f4056bbee2eb_
   > [Nivdia graphics passthrough video](https://www.youtube.com/watch?v=-Us8KPOhOCY)
   
   I have already installed NVIDIA drivers for jellyfin lxc in proxmox host, so I just installed the nvidia drivers in container using below commands:
   
   ```
    wget https://us.download.nvidia.com/XFree86/Linux-x86_64/535.146.02/NVIDIA-Linux-x86_64-{Latest-version}.run
   ```
    you can get the NVIDIA latest driver url from here : https://www.nvidia.com/Download/index.aspx
	
   once downloaded install the drivers (The --no-kernel-module module is used as already there's kernel installed ):
   ```
   chmod +x NVIDIA-Linux-x86_64-{Latest-version}.run
   
   ./NVIDIA-Linux-x86_64-{Latest-version}.run ----no-kernel-module
   ```
   
   once installed make sure the drivers are installed by running:
   ```
   nvidia-smi -l 10
   ```
   
10. Now we need to pass through the hardware acceleration to docker container by following the immich guide:
    
	```
	https://immich.app/docs/features/hardware-transcoding
    ```
	
	In order to have the driver passed to container, we need to have nvidia container toolkit installed using the following commands in the container:
	(https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/latest/install-guide.html)
	
	```
	curl -fsSL https://nvidia.github.io/libnvidia-container/gpgkey | sudo gpg --dearmor -o /usr/share/keyrings/nvidia-container-toolkit-keyring.gpg \
    && curl -s -L https://nvidia.github.io/libnvidia-container/stable/deb/nvidia-container-toolkit.list | \
    sed 's#deb https://#deb [signed-by=/usr/share/keyrings/nvidia-container-toolkit-keyring.gpg] https://#g' | \
    sudo tee /etc/apt/sources.list.d/nvidia-container-toolkit.list
	
	
	
	sed -i -e '/experimental/ s/^#//g' /etc/apt/sources.list.d/nvidia-container-toolkit.list
	
	
	
	sudo apt-get update
	
	
	sudo apt-get install -y nvidia-container-toolkit
	
	
	```
	
	Also one other step I needed to do to make sure the docker compose works with hardware-transcoding is to modify nvidia config using below commands in container:
	
	Edit /etc/nvidia-container-runtime/config.toml and disable cgroups:
	```
      
	  vi /etc/nvidia-container-runtime/config.toml
	  
	  set no-cgroups = true in the config.toml file, save and reboot the container. Now run the docker compose for immich.

	```
	
	
	
	References:
	
	https://immich.app/docs/install/script
	
	https://forum.proxmox.com/threads/how-to-automatically-mount-a-smb-share-on-an-lxc.118220/

    https://immich.app/docs/features/hardware-transcoding
	
	https://www.reddit.com/r/Proxmox/comments/s0ud5y/comment/jl4lef2/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button
	
	<blockquote class="reddit-embed-bq" data-embed-height="348"><a href="https://www.reddit.com/r/Proxmox/comments/s0ud5y/comment/jl4lef2/">Comment</a><br> by<a href="https://www.reddit.com/user/Masterofchicken/">u/Masterofchicken</a> from discussion<a href="https://www.reddit.com/r/Proxmox/comments/s0ud5y/cgroups2_problem_with_nvidiacontainercli/"><no value=""></no></a><br> in<a href="https://www.reddit.com/r/Proxmox/">Proxmox</a></blockquote><script async="" src="https://embed.reddit.com/widgets.js" charset="UTF-8"></script>

**Comment by timdonovanuk** [comment]
Has anyone had luck passing through normal Intel iGPU through to Immich in docker? 

My unprivileged lxc can see /dev/dri fine:

```
root@lxc-immich:/opt/immich# ls -l /dev/dri
drwxr-xr-x 2 nobody nogroup       80 Mar 11 10:37 by-path
crw-rw---- 1 nobody nogroup 226,   0 Mar 11 10:37 card0
crw-rw---- 1 nobody nogroup 226, 128 Mar 11 10:37 renderD128
```

And I've modified hwaccel.transcoding.yml to be:

```
  quicksync:
    user: "0:226"
    privileged: true
    devices:
      - /dev/dri:/dev/dri
```

But the immich-microservices docker container still reports nothing at /dev/dri 🤷

**Reply by timdonovanuk** [reply]
Hmm think I fixed this, I had to do `chmod 666 /dev/dri/renderD128` on the host machine.
